### ISSUE TITLE
Error with setting tensors to use cpu in packed_padded_sequence when CUDA tensor is set as default

### ISSUE THREAD

**Issue by ParticularlyPythonicBS** [pytorch/pytorch#16542]
## 🐛 Bug
@ptrblck  helped me diagnose the issue

if Default tensors are set to gpu, setting length to cpu for pack_padded_sequence doesn't work
## To Reproduce

`import torch`
`torch.set_default_tensor_type(torch.cuda.FloatTensor)`
`lengths, _ = torch.sort(torch.randint(1, 100, (10,)), descending=True)`
`lengths = torch.as_tensor(lengths.cpu(), dtype=torch.int64)`
`lengths.type()`
this results in a torch.cuda.LongTensor instead of a torch.Longtensor

This is due to the following line in RNN.py"
[line](https://github.com/pytorch/pytorch/blob/21193bf12314f9d1be5ea36cbce3162c54b35225/torch/nn/utils/rnn.py#L195)

## Expected behavior
Expected behaviour will be to accept the move to cpu.

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.13.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 410.79
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] cuda100                   1.0                           0    pytorch
[conda] mkl                       2019.1                      144  
[conda] mkl-service               1.1.2            py37he904b0f_5  
[conda] mkl_fft                   1.0.6            py37hd81dba3_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.0.0           py3.7_cuda10.0.130_cudnn7.4.1_1  [cuda100]  pytorch
[conda] torchvision               0.2.1                      py_2    pytorch

**Comment by ptrblck** [comment]
Thanks for posting this issue!

Based on the error message it seems `lengths` should always be a CPU tensor.
If that's the case, a simple fix could be just changing:
```python
lengths = torch.as_tensor(lengths, dtype=torch.int64)
```
to
```python
lengths = torch.as_tensor(lengths, dtype=torch.int64, device='cpu')
```

Let me know, if that's the case and I can create the PR.

Best,
ptrblck

**Comment by ParticularlyPythonicBS** [comment]
Yes, lengths should always be a CPU tensor, so the fix you proposed makes sense.

Best,
ParticularlyPythonicBS

**Comment by zou3519** [comment]
@ptrblck yes, that looks like a good fix to me, please send a patch when you get the chance. Thank you for surfacing the issue, @ParticularlyPythonicBS

**Comment by darcyYe** [comment]
Hi all!

As the issue still exists, is there any way to avoid it?

I am using following lines of code to put my model on GPU for training:

```
torch.cuda.set_device(0)
torch.set_default_tensor_type('torch.cuda.FloatTensor')
```

I can run forward pass of my model on cpu because it would no take a long time. Without using 'pack_padded_sequence' to training the model with batch, it may take long time to test some new models.

Is there alternative for such an issue?

Best,
Darcy

**Comment by ParticularlyPythonicBS** [comment]
https://discuss.pytorch.org/t/error-with-lengths-in-pack-padded-sequence/35517?u=particularlypythonic

**Comment by darcyYe** [comment]
> 
> 
> https://discuss.pytorch.org/t/error-with-lengths-in-pack-padded-sequence/35517?u=particularlypythonic

Ah, thanks so much!

It works after rewrite some modules!

**Comment by VinACE** [comment]
The cpu method did not fix Iam trying  lengths =  torch.as_tensor(lengths, dtype=torch.int64)

**Comment by futureisatyourhand** [comment]
> The cpu method did not fix Iam trying lengths = torch.as_tensor(lengths, dtype=torch.int64)



> The cpu method did not fix Iam trying lengths = torch.as_tensor(lengths, dtype=torch.int64)

Hello, I still have an error according to your method. How can I solve it?
error：caption = pack_padded_sequence(caption, lengths, batch_first=True)[0]
  File "/home/liqian19/anaconda/lib/python3.6/site-packages/torch/nn/utils/rnn.py", line 148, in pack_padded_sequence
    return PackedSequence(torch._C._VariableFunctions._pack_padded_sequence(input, lengths, batch_first))
RuntimeError: 'lengths' argument should be a 1D CPU int64 tensor

**Comment by ptrblck** [comment]
@futureisatyourhand Did you also set the default tensor to a CUDATensor?

I'll follow up on the PR, as I've thought it was merged already. Sorry for the delay.

**Comment by sergicastellasape** [comment]
Hi! I see this issue is still open, is there any update on it? I see that the new 1.4.0 update still contains the same 'problematic' line of code and I've encountered this same bug recently and the only workaround I could manage to make work was the hotfix @ptrblck posted on the PyTorch forum; the fix here didn't work for me!

Thanks

**Comment by ParticularlyPythonicBS** [comment]
Hi! Yeah the issue still doesn't seem to have a complete fix.
The most seemless way I found to fix the problem is to trap the packing code with cpu as default like so:
```torch.set_default_tensor_type(torch.FloatTensor)
        padded_track_seq = pack_padded_sequence(sorted_tracks, sorted_n_tracks, batch_first=True, enforce_sorted=True)
        if self.device == torch.device("cuda"):
            torch.set_default_tensor_type(torch.cuda.FloatTensor)
        padded_track_seq.to(self.device)```

**Comment by ankit-kothari** [comment]
Hey, any updates on this issue. I am getting the same error and I tried working with the hotfix. I am not getting the result. Thanks

**Comment by futureisatyourhand** [comment]
> @futureisatyourhand Did you also set the default tensor to a CUDATensor?
> 
> I'll follow up on the PR, as I've thought it was merged already. Sorry for the delay.

Nothing. I'm setting it to CUDA by default.

**Comment by ntuLC** [comment]
It is somehow sad to see that the PR #16636 is still not have been accepted :(

I'm using the fix mentioned by @ParticularlyPythonicBS. Works nice for me but I wonder if this will affect the executing speed of the script?

**Comment by ntuLC** [comment]
It is quite sad to see that the PR #16636 is still not have been accepted :(

I'm using the fix mentioned by @ParticularlyPythonicBS. Works nice for me but I wonder if this will affect the executing speed of the script?

**Comment by ankit-kothari** [comment]
Could you guys please help me implement it in the current form, I am not able to get it to work from the solution provided. 

Thanks 
Ankit
